text/0068-entity-tags.md

```diff
@@ -0,0 +1,296 @@
+```
+  resource.policyTags has "write" && // (1)
+  principal.authTags has "write" &&  // (2)
+  resource.policyTags["write"].containsAny(principal.authTags["write"]) // (3)
+```
+After the subexpression (1), `resource.policyTags.write` should be added to the current capability set. After subexpression (2), `principal.authTags.write` should be added to it. Finally, when validating subexpression (3), the expression `resource.policyTags["write"]` will be considered valid since `resource.policyTags.write` is in the current capability set and it will be given type `Set<String>`, as `resource.policyTags` has type `Tags<Set<String>>`. The expression `principal.authTags["write"]` is handled similarly. If either of the `has` subexpressions (1) or (2) were omitted, subexpression (3) would not validate due to the missing capability set entries.
+
+#### Limited operations
+
+Tags' second-class status means that expressions of types `Tags<T>` are only permitted as part of `has` and projection expressions (i.e., using `.` or `[]`). The validator should disallow these expressions in other contexts, e.g., the following should be disallowed:
+```
+resource.policyTags == principal.authTags
+[resource.policyTags, principal.authTags]
+(if true then principal.authTags else principal.authTags).write
+```
+(Note that the last case is disallowed because `principal.authTags` and `principal.authTags` are not directly part of a `.` sub-expression -- they are returned from the `if` first.)
+
+### Validating and parsing entities
+
+The Cedar authorizer's `is_authorized` function can be asked to validate that entities in a request are consistent with a provided schema. Extending validation to work with tags is straightforward. The type rule is basically thus:
+```
+v1: T ... vn: T
+---------------------------------
+{ f1: v1, ..., fn: vn } : Tags<T>
+```
+In particular, when asked to determine if a record has type `Tags<T>`, we confirm that all values in the record have type `T`. By the nature of restrictions on schemas, the validator will only ever consider `Tags<T>` types associated with entity attributes.
+
+Similarly, schema-based parsing considers schemas when parsing in entities, and it can confirm when parsing that attributes labeled as `Record` in the JSON but defined as `Tags<T>` in the schema have the appropriate shape.
+
+### Permissive Validation
+
+Permissive validation supports subtyping, which we extend so as to allow `Tags` types to be treated co-variantly:
+```
+Tags<T> <: Tags<U>    iff T <: U
+```
+We might consider adding rules that permit subtyping between records and tags, but doing so introduces some complications with width subtyping.
+
+We could choose to grant `Tags` first-class status under permissive validation. For example, we could allow `Tags` values to appear anywhere, e.g., in `Set`s or records, and we could allow equality between `Tags` values. We choose not to do this for now, but may revisit in the future.
+
+## Drawbacks
```

Comment:
> the same expressive power proposed in this RFC (or something of similar power, and sufficient for the motivating use cases), but exposed to users at the syntax level 

We need to extend _entity_ types for feasible analysis. And that is precisely what we are doing at a syntactic level. If you are suggesting that we extend record types but encode them as entities under the covers, perhaps just for analysis, then _maybe_ that could work, but it would add a fair bit of complexity to the analysis code (especially for generating counterexamples). It would also be confusing/disappointing to users that default-typed records are more restricted than normal records. E.g., adding a default type to an existing record may cause validation to break, and could require a data-model rethink.

> maybe records are allowed to have some kind of default type, but if they do they are not allowed to have other attributes

This will not help the analyzability problem, since records with unknown/arbitrary attributes could appear in equality operations, which can't be feasibly encoded. (Extending an _entity_ with unknown attributes works because we never test entity _contents_ for equality, only entity references.)

> I think there's some available design space here and I think we should explore it as a community. 

Have you read RFCs 27 and 66 closely? These explore what I see as a big chunk of the available design space, but they have proved to be inadequate from an analysis perspective. You may be right that there are other options, but we _have_ thought seriously and hard about this problem for many months now, and we haven't come up with anything. Please present your ideas here!

Reply:
My concern in this thread is sufficiently addressed by the new syntax proposed in the latest revision.  I'm happy to see a syntax that looks a lot more like an extension of the existing `entity` type, rather than an entirely new third kind of key-value mapping.